quinn-proto/src/connection/mod.rs

```diff
@@ -889,23 +889,23 @@ where
         Some(id)
     }
 
-    /// Read from the given recv stream, in undefined order
+    /// Read from the given recv stream
     ///
-    /// While stream data is typically processed by applications in-order, unordered reads improve
-    /// performance when packet loss occurs and data cannot be retransmitted before the flow control
-    /// window is filled. When in-order delivery is required, the sibling `read()` or `read_chunk[s]()`
-    /// methods should be used.
+    /// If `Ok`, the return value contains the bytes and their offset in the stream. For ordered
+    /// reads, the offset could be ignored, since it provides no new information.
```

Comment:
Let's actually define "ordered" here, e.g. as "having an offset exactly equal to the previously returned offset plus the previously returned bytes' length"

Reply:
Revised the comments here. Related question: we used to have some docstrings stating that unordered reads could yield chunks in any order and might yield duplicate data, but that's no longer true, right? As far a I understand it, we always yield chunks in order and we now deduplicate them on insertion when the assembler is in unordered mode because we need to do flow control.

Actually, I wonder if what we do is correct and the most simple thing we could do here. In my understanding, the assembler deduplicates on read when in ordered mode whereas we deduplicate on ingest in unordered mode. Could we deduplicate on read always to simplify the code? The deduplication in unordered mode is currently a little hairy. Or, vice versa, isn't deduplication on read too lazy?

Reply:
> we always yield chunks in order

This is only true if we're not missing any chunks. Unordered reads can return out-of-order chunks due to network errors or weird peer behavior, whereas ordered reads just block in those cases.

> Could we deduplicate on read always?

Consider this situation:
1. Peer sends offset 5 on a new stream
2. Application reads offset 5 in unordered mode
3. Peer sends offset 5 again

We must not issue flow control credit for the same offset twice. However, we've already evicted the first copy of that data from the buffer. In ordered mode, this is trivial because we've issued flow control credit for an offset iff it's less than `self.bytes_read`. In unordered mode, the ranges consumed by the application may be sparse, so we need the `RangeSet`.

> Or, vice versa, isn't deduplication on read too lazy?

We must also avoid spending unbounded buffer space on data that we've already issued flow control credit for, so we must deduplicate on insert rather than waiting an unbounded duration for the application to get around to reading. I think you're right that we currently have a bug here: in ordered mode, offsets prior to `self.bytes_read` should be immediately dropped, but currently are not. In unordered mode we're behaving correctly.

Reply:
> This is only true if we're not missing any chunks. Unordered reads can return out-of-order chunks due to network errors or weird peer behavior, whereas ordered reads just block in those cases.

Ah, fair.

> In unordered mode, the ranges consumed by the application may be sparse, so we need the RangeSet.
> 
> It is true that we can now guarantee not returning duplicate data, as an incidental benefit of the bookkeeping required to avoid issuing duplicate flow control credit.

Right.

> I think you're right that we currently have a bug here: in ordered mode, offsets prior to self.bytes_read should be immediately dropped, but currently are not. In unordered mode we're behaving correctly.

Good -- that should be a straightforward fix.